api/controllers/AssignmentController.js

```diff
@@ -119,14 +98,60 @@ module.exports = class AssignmentController extends classMixin(OrderedDataContro
       )
     }
 
-    if (judgement.grade > 0 && !parameterizedAssignment.dataValues.solved) {
-      await parameterizedAssignment.update({ solved: true })
-    }
-
     return res.status(200).send(
       {
         success: true, judgement
       }
     )
   }
+
+  async processAsyncJudge (req, res, aux, assignment, solution, parameterizedAssignment) {
+    let judgement = { grade: 0, msg: 'Wrong! Please try again.' }
+    const judge = new assignment.Judge(assignment.judge, req.user)
+
+    res.status(202).send({ code: 202, judgement: { grade: 0, msg: 'Solution is being processed. Please wait!' } })
+
+    try {
+      judgement = await judge.judge(aux, req.user, assignment.Judge, solution)
+      await this.updateSolution(req, parameterizedAssignment, solution)
+      await this.updateSolved(judgement, parameterizedAssignment)
+      appEmitter.emit('solution', { ...judgement })
+    } catch (err) {
+      logger.error(err)
+      judgement = { grade: 0, msg: err.message }
+
+      appEmitter.emit('solution', { ...judgement })
```

Comment:
This event emission should include details about the assignment, not just the solution. Imagine a user who has two tabs open and submits two solutions almost simultaneously. Then when the first of the two returns, it will cause the second one to incorrectly update the front-end according to the solution of the first.

Reply:
Ok. Nevertheless, you cannot say for sure that the user will stay on assignment's page. She could navigate to another page and have in mind that she will get her notification about the exercise. So assignment details would help only for the message and not as a mean of check. Right ?

Reply:
Like putting the title of the assignment in front of the message

Reply:
Sounds OK, yes. Put it as a title in bold for example.